avocado/utils/vmimage.py

```diff
@@ -0,0 +1,251 @@
+            name = 'CentOS'
+
+        super(CentOS, self).__init__(name=name,
+                                     major=major,
+                                     minor=minor,
+                                     build=build,
+                                     arch=arch,
+                                     checksum=checksum,
+                                     algorithm=algorithm,
+                                     cache_dir=cache_dir,
+                                     url=url)
+
+
+class Fedora(Image):
+    def __init__(self, name=None, major=None, minor=None, build=None,
+                 arch=None, checksum=None, algorithm=None, cache_dir=None):
+        """
+        Fedora Qcow2 Cloud Base image
+        """
+        required_args = ['major', 'minor', 'build', 'arch']
+        for arg in required_args:
+            if eval(arg) is None:
+                raise VMImageError('Required args: %s' % ','.join(required_args))
+
+        url = ('https://download.fedoraproject.org/'
+               'pub/fedora/linux/releases/{major}/CloudImages/{arch}/images/'
+               'Fedora-Cloud-Base-{major}-{minor}.{build}.{arch}.qcow2')
+        url = url.format(major=major, minor=minor, build=build, arch=arch)
+
+        if name is None:
+            name = 'Fedora'
+
+        super(Fedora, self).__init__(name=name,
+                                     major=major,
+                                     minor=minor,
+                                     build=build,
+                                     arch=arch,
+                                     checksum=checksum,
+                                     algorithm=algorithm,
+                                     cache_dir=cache_dir,
+                                     url=url)
+
+
+class Ubuntu(Image):
+    def __init__(self, name=None, major=None, minor=None, build=None,
+                 arch=None, checksum=None, algorithm=None, cache_dir=None):
+        """
+        Ubuntu Qcow2 Server Cloud image
+        """
+        required_args = ['major', 'minor', 'arch']
+        for arg in required_args:
+            if eval(arg) is None:
+                raise VMImageError('Required args: %s' % ','.join(required_args))
+
+        url = ('https://cloud-images.ubuntu.com/'
+               'releases/{major}.{minor}/release/'
+               'ubuntu-{major}.{minor}-server-cloudimg-{arch}.img')
+        url = url.format(major=major, minor=minor, arch=arch)
+
+        if name is None:
+            name = 'Ubuntu'
+
+        super(Ubuntu, self).__init__(name=name,
+                                     major=major,
+                                     minor=minor,
+                                     build=build,
+                                     arch=arch,
+                                     checksum=checksum,
+                                     algorithm=algorithm,
+                                     cache_dir=cache_dir,
+                                     url=url)
```

Comment:
How would you expect a user to interact with this library?  As a user, I'd be happy doing something like:

```
>>> from avocado.utils import vmimage

>>> centos = vmimage.get('centos', major='7')
>>> centos.path
"/var/lib/avocado/data/vmimages/CentOS-7-x86_64-GenericCloud-1707.qcow2.fe1a93ed12"
>>> centos.arch
"x86_64"

>>> centos_s390x = vmimage.get('centos', major='7', arch='s390x')
...
VMImageProviderError: Provider CentosImageProvider did not find a matching image for "major=7" + "arch=s390x"

>>> reactos = vmimage.get('reactos', version='0.4.5')
...
VMImageProviderError: no provider for images of name 'reactos' are registered

>>> vmimage.get_providers_list()
[<CentosImageProvider name="centos">, <FedoraImageProvider name="fedora">, ...]
```

Reply:
👍 to backends vs. classes....